README.md
# vscode-wordcount-cjk

This is a more powerful word count extension for VSCode, which supports the CJK character counting.

## Features

In current version, the following things are counted:

1. Total characters in a document.
2. Non-whitespace characters.
3. English words.
4. Non-ASCII characters.
5. CJK characters.

In current version, the following format is supported:

1. Markdown file.
2. Plain text file.

For other files that are not supported, such as ReStructuredText, a command `Word Count` is offered. The command only calculate once. For keep tracking the word count, you can use the `Word Count Activate` command, and the `Word Count Deactivate` to stop tracking.

A status bar item is added, and full statistics are added as a tooltip of the status bar item.

In current version, all CJK characters are counted as Chinese characters. Since I have little knowledge of
Japanese and Korean, I don't know how to count them. If someone knows, please
tell me in an issue.

## Extension Settings

* `wordcount_cjk.statusBarTextTemplate`: Customize the status bar item text.
* `wordcount_cjk.statusBarTooltipTemplate`: Customize the status bar item tooltip.
* `wordcount_cjk.regexWordChar`: The regular expression used to test if a char is a word char.
* `wordcount_cjk.regexASCIIChar`: The regular expression used to test if a char is a ASCII char.
* `wordcount_cjk.regexWhitespaceChar`: The regular expression used to test if a char is a whitespace.
* `wordcount_cjk.activateLanguages`: The languages that activate the extension.

in `wordcount_cjk.statusBarTextTemplate` and `wordcount_cjk.statusBarTooltipTemplate`, the following placeholder could be used:

1. `cjk`: The number of CJK characters.
2. `ascii`: The number of ASCII characters.
3. `whitespace`: The number of whitespace characters.
4. `en_words`: The number of english words.
5. `total`: The total number of characters.

Since version 1.1, calculation could be done in format string, so one can do this: `${total - cjk} chars`. The expression should
be a valid JavaScript expresion.

### Word char

In default settings, word char is tested by regular expression `\w`, which is equivalent to `[A-Za-z0-9_]`.

### ASCII char

In default settings, word char is tested by regular expression `[\u0000-\u00FF]`, which contains:

1. `0000-007F`: C0 Controls and Basic Latin (ASCII)
2. `0080-00FF`: C1 Controls and Latin-1 Supplement (Extended ASCII)

Perhaps someone want to limit the counter with basic ASCII, then config the `wordcount_cjk.regexASCIIChar` to
`[\u0000-\u007F]`.

### Whitespace char

In default settings, word char is tested by regular expression `\s`, which is equivalent to
`[ \f\n\r\t\v\u00a0\u1680\u180e\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff]`. That means
unicode whitespace is also counted. If you don't what that, change the regular expression
`wordcount_cjk.regexWhitespaceChar`.

### CJK char

**NOTE**: This is not configurable.

The built-in regular expression is `[\u4E00-\u9FA5\uF900-\uFA2D]`, which contains:

* 4E00-9FFF: CJK Unified Ideographs
* F900-FAFF: CJK Compatibility Ideographs

if someone other than Chinese use this extension, I will try to seperate the Chinese,
Japanese and Korean characters counters. For now, they are mixed.

## TODO

1. None

If you have any requested feature, open an issue!

**Enjoy!**

CHANGELOG.md
# Change Log
All notable changes to the "wordcount-cjk" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [1.2.0] - 2018-06-01
### Enhancement
- Add `activateLanguages` options.
- Add `Word Count Activate` and `Word Count Deactivate` command

### Remove
- Remove the activate method by `.editorconfig`.

## [1.1.0] - 2018-01-28
### Enhancement
- Format string supports calculation. [Issue #2](https://github.com/holmescn/vscode-wordcount-cjk/issues/2)

### Fix
- Add a timer to update count, which fix the first issue of delay count.

## [1.0.1] - 2018-01-22
### Added
- Add a new activate method through `.editorconfig`. [Issue #1](https://github.com/holmescn/vscode-wordcount-cjk/issues/1)

## [Unreleased]
- Initial release


src/extension.ts
'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { WordCounter } from './WordCounter';
import { WordCountController } from './WordCountController';

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
    const configuration = vscode.workspace.getConfiguration("wordcount_cjk");
    let counter = new WordCounter(configuration);
    let controller = new WordCountController(configuration, counter);
    let command1 = vscode.commands.registerCommand('extension.wordcount', () => {
        // This command is used to activate the extension when
        // edit non-standard type files.
        controller.update(/* force = */true);
    });
    let command2 = vscode.commands.registerCommand('extension.wordcountActivate', () => {
        // This command is used to activate the extension when
        // edit non-standard type files.
        controller.activate();
    });
    let command3 = vscode.commands.registerCommand('extension.wordcountDeactivate', () => {
        // This command is used to deactivate the extension
        controller.deactivate();
    });

    context.subscriptions.push(command1);
    context.subscriptions.push(command2);
    context.subscriptions.push(command3);
    context.subscriptions.push(controller);
}

// this method is called when your extension is deactivated
export function deactivate() {
}


src/WordCountController.ts
// Import the module and reference it with the alias vscode in your code below
import { window, StatusBarItem, StatusBarAlignment, Disposable, WorkspaceConfiguration } from 'vscode';
import { WordCounter } from './WordCounter';

/**
 * The controller of word count
 */
export class WordCountController {
    private _wordCounter: WordCounter;
    private _statusBarItem: StatusBarItem;
    private _disposable: Disposable;
    private _delayUpdateTimer: any;
    private _isActive: boolean;
    private readonly _statusBarTextTemplate: string;
    private readonly _statusBarTooltipTemplate: string;
    private readonly _activateLanguages: Array<string>;

    constructor(configuration: WorkspaceConfiguration, wordCounter: WordCounter) {
        this._isActive = false;
        this._wordCounter = wordCounter;
        this._statusBarTextTemplate = configuration.get<string>("statusBarTextTemplate");
        this._statusBarTooltipTemplate = configuration.get<string>("statusBarTooltipTemplate");
        this._activateLanguages = configuration.get<Array<string>>("activateLanguages");

        this._statusBarItem = window.createStatusBarItem(StatusBarAlignment.Left);

        // subscribe to selection change and editor activation events
        let subscriptions: Disposable[] = [];
        window.onDidChangeTextEditorSelection(this._onEvent, this, subscriptions);
        window.onDidChangeActiveTextEditor(this._onEvent, this, subscriptions);

        // create a combined disposable from both event subscriptions
        this._disposable = Disposable.from(...subscriptions);
    }

    public update(force: boolean = false) {
        let editor = window.activeTextEditor;
        if (!editor) {
            this._statusBarItem.hide();
            return;
        }

        let doc = editor.document;

        if (force || this.shouldActivate(doc.languageId) || this._isActive) {
            // Update word count.
            let text = doc.getText();
            this._wordCounter.count(text);

            // Update the status bar
            try {
                this._statusBarItem.text = this._wordCounter.format(this._statusBarTextTemplate);
                this._statusBarItem.tooltip = this._wordCounter.format(this._statusBarTooltipTemplate);
                this._statusBarItem.show();
            } catch (e) {
                window.showErrorMessage('Something is wrong when update status bar');
            }
        } else {
            this._statusBarItem.hide();
        }
    }

    private shouldActivate(languageId: string): boolean {
        for(var l of this._activateLanguages) {
            if (l === languageId) {
                return true;
            }
        }

        return false;
    }

    private _onEvent() {
        this.update();

        if (this._delayUpdateTimer) {
            clearTimeout(this._delayUpdateTimer);
        }

        this._delayUpdateTimer = setTimeout(() => {
            this._delayUpdateTimer = null;
            this.update();
        }, 500);
    }

    /**
     * This makes the class disposable.
     */
    dispose() {
        this._statusBarItem.dispose();
        this._disposable.dispose();
    }

    activate() {
        this._isActive = true;
        this.update();
    }

    deactivate() {
        this._isActive = false;
        this._statusBarItem.hide();
    }
}

src/WordCounter.ts
import { WorkspaceConfiguration } from "vscode";

/**
 * The main class to provide counter functionality.
 */
export class WordCounter {

    /** 中文字数 */
    private _nChineseChars: number = 0;
    /** 非 ASCII 字符娄 */
    private _nASCIIChars: number = 0;
    /** 英文单词数 */
    private _nEnglishWords: number = 0;
    /** 非空白字符娄 */
    private _nWhitespaceChars: number = 0;
    /** 总字符数 */
    private _nTotalChars: number = 0;
    /** 格式化函数 */
    private _replaceFuncs: { [key: string]: Function }

    private readonly _regexWordChar: RegExp;
    private readonly _regexASCIIChar: RegExp;
    private readonly _regexWhitespaceChar: RegExp;
    private readonly _regexFormatReplace: RegExp;

    constructor(configuration: WorkspaceConfiguration) {
        this._regexWordChar = new RegExp(configuration.get<string>("regexWordChar"));
        this._regexASCIIChar = new RegExp(configuration.get<string>("regexASCIIChar"));
        this._regexWhitespaceChar = new RegExp(configuration.get<string>("regexWhitespaceChar"));
        this._regexFormatReplace = /\$\{([^}]+)\}/g;
        this._replaceFuncs = {};
    }

    public count(text: string) {
        this._resetCounters();

        let inWord = false;
        for (let index = 0; index < text.length; ++index) {
            // Get a char
            const ch = text.charAt(index);

            this._countChineseChar(ch);
            this._countASCIIChar(ch);
            inWord = this._countEnglishWord(ch, inWord);
            this._countWhitespaceChar(ch);
        }
        this._countEnglishWord(' ', inWord);

        this._nTotalChars = text.length;
    }

    public format(fmt: string) : string {
        const _this = this;
        const cjk = this._nChineseChars;
        const ascii = this._nASCIIChars;
        const whitespace = this._nWhitespaceChars;
        const en_words = this._nEnglishWords;
        const total = this._nTotalChars;

        return fmt.replace(this._regexFormatReplace, (matches, expr) => {
            if (!_this._replaceFuncs[matches]) {
                _this._replaceFuncs[matches] = _this._compileExpiession(expr);
            }

            const f = _this._replaceFuncs[matches];
            return f(cjk, ascii, whitespace, en_words, total);
        });
    }

    private _compileExpiession(expr: string) {
        const f = new Function('cjk', 'ascii', 'whitespace', 'en_words', 'total', `return ${expr};`);

        try {
            f(0, 0, 0, 0, 0);
        } catch (e) {
            return new Function('cjk', 'ascii', 'whitespace', 'en_words', 'total', `return '无效表达式: ${expr}';`);
        }

        return f;
    }

    /**
     * Justify if a char is ASCII character.
     *
     * 0000-007F: C0 Controls and Basic Latin (ASCII)
     * 0080-00FF: C1 Controls and Latin-1 Supplement (Extended ASCII)
     *
     * Reference:
     * http://houfeng0923.iteye.com/blog/1035321 (Chinese)
     * https://en.wikipedia.org/wiki/Basic_Latin_(Unicode_block)
     * https://en.wikipedia.org/wiki/Latin-1_Supplement_(Unicode_block)
     *
     * @todo Current ASCII contains ASCII and Extended ASCII.
     * This sould be configurable in the future.
     *
     * @param ch char to be tested.
     */
    private _countASCIIChar(ch: string) {
        if (this._regexASCIIChar.test(ch)) {
            this._nASCIIChars += 1;
        }
    }

    /**
     * Justify if a char is Chinese character.
     *
     * 4E00-9FFF: CJK Unified Ideographs
     * F900-FAFF: CJK Compatibility Ideographs
     *
     * Reference:
     * http://houfeng0923.iteye.com/blog/1035321 (Chinese)
     * https://en.wikipedia.org/wiki/CJK_Unified_Ideographs
     * https://en.wikipedia.org/wiki/CJK_Compatibility_Ideographs
     *
     * @todo Refine to contains only Chinese Chars.
     *
     * @param ch Char to be tested.
     */
    private _countChineseChar(ch: string) {
        // Count chinese Chars
        const regexChineseChar = /[\u4E00-\u9FA5\uF900-\uFA2D]/;
        if (regexChineseChar.test(ch)) {
            this._nChineseChars += 1;
        }
    }

    /**
     * Test if a char is a word character.
     *
     * The regular expression \w rule is used.
     * In Javascript it matches any alphanumeric character including the underscore.
     * Equivalent to [A-Za-z0-9_].
     *
     * @todo the word means should be configurable.
     *
     * @param ch Char to be tested
     */
    private _countEnglishWord(ch: string, inWord: boolean) : boolean {
        if (this._regexWordChar.test(ch)) {
            return true;
        } else {
            if (inWord) {
                this._nEnglishWords += 1;
            }
            return false;
        }
    }

    /**
     * Test if a char is a white space.
     *
     * The regular expression \s rule is used.
     * In Javascript it matches a single white space character, including space,
     * tab, form feed, line feed.
     * Equivalent to [ \f\n\r\t\v\u00a0\u1680\u180e\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff]
     *
     * @param ch Char to be tested.
     */
    private _countWhitespaceChar(ch: string) {
        if (this._regexWhitespaceChar.test(ch)) {
            this._nWhitespaceChars += 1;
        }
    }

    private _resetCounters() {
        this._nChineseChars = 0;
        this._nASCIIChars = 0;
        this._nEnglishWords = 0;
        this._nWhitespaceChars = 0;
        this._nTotalChars = 0;
    }
}